### PR TITLE
Added installation utilities

### LINKED ISSUE #72: Install Utils class
As a user, I'd want an easy way of installing the software needed if I'm missing some.

### PR DESCRIPTION
**Pull request by veikkahonkanen**
Resolves #72

### CODE REVIEW

File: src/classes/install-sw-refs.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  public static async getInstallRef(installUtil : string, software : string) : Promise<string> {
+
+    switch (software) {
+      case Software.NodeJs:
+        if (installUtil == "brew") {
+          return "node";
+        } else {
+          return "nodejs";
+        }
```
Comment (eetupur) [inline src/classes/install-sw-refs.ts]: could use a shorter syntax on many of these
```suggestion
       return (installUtil === "brew") ? "node" : "nodejs";
```

File: src/classes/install-sw-refs.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+      case Software.GitHub:
+        return "gh";
+
+      case Software.GitCLI:
+        if (installUtil == "sudo apt") {
```
Comment (eetupur) [inline src/classes/install-sw-refs.ts:26]: same here

File: src/classes/install-sw-refs.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+          return "git-all";
+        } else {
+          return "git";
+        }
```
Comment (eetupur) [inline src/classes/install-sw-refs.ts]: same here

File: src/classes/install-sw-refs.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+      case Software.Maven:
+        if (installUtil == "brew") {
+          return "maven@3.5";
+        } else {
+          return "maven";
+        }
```
Comment (eetupur) [inline src/classes/install-sw-refs.ts]: same here

File: src/classes/install-utils.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        result = execSync(`which ${software}`).toString();
+      } catch (err) {
+        throw new Error(`Error when checking software ${software}: ${err}`);
+      }
+      
+      if(result.search(/not found/) == -1) {
+        return true;
+      } else {
+        return false;
+      }
```
Comment (eetupur) [inline src/classes/install-utils.ts]: same here

File: src/classes/install-utils.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        result = execSync(`${software} --version`).toString();
+      } catch (err) {
+        throw new Error(`Error when checking software ${software}: ${err}`);
+      }
+      
+      if(result.search(/is not recognized/) == -1) {
+        return true;
+      } else {
+        return false;
+      }
```
Comment (eetupur) [inline src/classes/install-utils.ts]: same here

**Review by eetupur — COMMENTED** [review]
found cases where better syntax could be used

**Review by eetupur — APPROVED** [review]

**Review by eetupur — APPROVED** [review]

File: src/classes/install-sw-refs.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
+import { Software } from "../interfaces/types";
+
+export class InstallSwRefs {
```
Comment (jarinykanen) [inline src/classes/install-sw-refs.ts:6]: Add tsdocs

File: src/classes/install-sw-refs.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * Get Install Reference for Installation Utils class
+   * 
+   * @param software is the software in question
+   * 
+   * @returns installation reference used in bash to install the software
+   */
+  public static async getInstallRef(installUtil : string, software : string) : Promise<string> {
```
Comment (jarinykanen) [inline src/classes/install-sw-refs.ts]:
```suggestion
  public static async getInstallRef(installUtil: string, software: string): Promise<string> {
```

Comment (jarinykanen) [reply src/classes/install-sw-refs.ts]: Add also missing tsdocs

File: src/classes/install-sw-refs.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  /**
+   * Get bash references for Installation Utils class
+   * 
+   * @param software is the software in question
+   * 
+   * @returns bash reference for the software
+   */
+  public static async getBashRef(software : string) : Promise<string> {
```
Comment (jarinykanen) [inline src/classes/install-sw-refs.ts]:
```suggestion
  public static async getBashRef(software: string): Promise<string> {
```

File: src/classes/install-utils.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import { execSync } from "child_process";
+import OsUtils from "./os-utils"
+import { OperatingSystems, CommandNames } from "../interfaces/types";
+import { InstallSwRefs } from "./install-sw-refs";
+
+export class InstallUtils {
```
Comment (jarinykanen) [inline src/classes/install-utils.ts:9]: Add docs

File: src/classes/install-utils.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+  /**
+   * Install Homebrew command for MAC OS users
+   * 
+   * @returns Homebrew installation command that is run by the wizard
+   */
+  public static async installBrew() : Promise<string> {
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  public static async installBrew(): Promise<string> {
```

File: src/classes/install-utils.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    const os : string = await OsUtils.getOS();
+    if (os == OperatingSystems.MAC){
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
    const os : string = await OsUtils.getOS();

    if (os == OperatingSystems.MAC){
```

File: src/classes/install-utils.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  /**
+   * Install software command for bash
+   * 
+   * @param software is the desired software to be installed
+   * 
+   * @returns software installation command that is run by the wizard
+   */
+  public static async installSW(software : string) : Promise<string> {
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  public static async installSW(software: string): Promise<string> {
```

File: src/classes/install-utils.ts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    } else {
+      try {
+        const str = `${software} --version`;
+        result = execSync(str).toString();
+      } catch (err) {
+        throw new Error(`Error when checking software ${software}: ${err}`);
+      }
+      return (result.search(/is not recognized/) == -1);
+    }
+  }
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  public static async isInstalled(software: string): Promise<boolean> {
    const bashRef = await InstallSwRefs.getBashRef(software);

    if(bashRef == "brew"){
      try {
        const str = `which ${bashRef}`;
        const result = execSync(str).toString();
        return (result.search(/not found/) == -1);
      } catch (err) {
        throw new Error(`Error when checking software ${bashRef}: ${err}`);
      }

    } else {
      try {
        const str = `${bashRef} --version`;
        const result = execSync(str).toString();
        return (result.search(/is not recognized/) == -1);
      } catch (err) {
        throw new Error(`Error when checking software ${bashRef}: ${err}`);
      }
    }
  }
```

File: src/interfaces/types.ts [hunk 15]
```diff
@@ -81,5 +81,16 @@ export enum OperatingSystems {
 export enum CommandNames {
   copy = "copy",
   remove = "remove",
-  removeDir = "remove directory"
+  removeDir = "remove directory",
+  installUtil = "installation utility"
+}
+
+export enum Software {
```
Comment (jarinykanen) [inline src/interfaces/types.ts:100]: Add tsdocs

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/classes/install-sw-refs.ts (hunk 7 above)
Comment (veikkahonkanen) [reply src/classes/install-sw-refs.ts:6]: added tsdocs

File: src/classes/install-sw-refs.ts (hunk 8 above)
Comment (veikkahonkanen) [reply src/classes/install-sw-refs.ts]: reformatted line 12, added tsdocs

File: src/classes/install-utils.ts (hunk 10 above)
Comment (veikkahonkanen) [reply src/classes/install-utils.ts:9]: added docs

File: src/interfaces/types.ts (hunk 15 above)
Comment (veikkahonkanen) [reply src/interfaces/types.ts:100]: added tsdocs

File: src/classes/install-utils.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    if(bashRef == "brew"){
+      try {
+        const str = `which ${bashRef}`;
+        const result = execSync(str).toString();
+        return (result.search(/not found/) == -1);
+
+      } catch (err) {
+        throw new Error(`Error when checking software ${software}: ${err}`);
+      }
+
```
Comment (eetupur) [inline src/classes/install-utils.ts]:
```suggestion
```

File: src/classes/install-utils.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        throw new Error(`Error when checking software ${software}: ${err}`);
+      }
+      
+
+    } else {
+      try {
+        const str = `${bashRef} --version`;
+        const result = execSync(str).toString();
+        return (result.search(/is not recognized/) == -1);
+
```
Comment (eetupur) [inline src/classes/install-utils.ts]:
```suggestion
```

**Review by eetupur — CHANGES_REQUESTED** [review]
found unnecessary lines

File: src/classes/install-sw-refs.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Get Install Reference for Installation Utils class
+   * 
+   * @param installUtil is the installation utility which needs its corresponding command
+   * 
+   * @param software is the software in question
+   * 
+   * @returns installation reference used in bash to install the software
+   */
+  public static async getInstallRef(installUtil: string, software: string): Promise<string> {
```
Comment (jarinykanen) [inline src/classes/install-sw-refs.ts]:
```suggestion
  /**
   * Get Install Reference for Installation Utils class
   * 
   * @param installUtil is the installation utility which needs its corresponding command
   * @param software is the software in question
   * @returns installation reference used in bash to install the software
   */
  public static async getInstallRef(installUtil: string, software: string): Promise<string> {
```

File: src/classes/install-sw-refs.ts [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    }
+  }
+
+  /**
+   * Get bash references for Installation Utils class
+   * 
+   * @param software is the software in question
+   * 
+   * @returns bash reference for the software
+   */
```
Comment (jarinykanen) [inline src/classes/install-sw-refs.ts]:
```suggestion
   * Get bash references for Installation Utils class
   * 
   * @param software is the software in question
   * @returns bash reference for the software
   */
```

File: src/classes/install-utils.ts [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+ */
+export class InstallUtils {
+
+  /**
+   * Install Homebrew command for MAC OS users
+   * 
+   * @returns Homebrew installation command that is run by the wizard
+   */
+  public static async installBrew(): Promise<string> {
+    const os : string = await OsUtils.getOS();
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
    const os: string = await OsUtils.getOS();
```

File: src/classes/install-utils.ts [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+   * @param software is the desired software to be installed
+   * 
+   * @returns software installation command that is run by the wizard
+   */
+  public static async installSW(software: string): Promise<string> {
+    const installUtil : string = await OsUtils.getCommand(CommandNames.installUtil);
+    const installRef : string = await InstallSwRefs.getInstallRef(installUtil, software);
+
+    return `${installUtil} install ${installRef}`;
+  }
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  /**
   * Install software command for bash
   * 
   * @param software is the desired software to be installed
   * @returns software installation command that is run by the wizard
   */
  public static async installSW(software: string): Promise<string> {
    const installUtil: string = await OsUtils.getCommand(CommandNames.installUtil);
    const installRef: string = await InstallSwRefs.getInstallRef(installUtil, software);

    return `${installUtil} install ${installRef}`;
  }
```

File: src/classes/install-utils.ts [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * Attempts to confirm a software's installation status
+   * 
+   * @param software is the software that is being checked
+   * 
+   * @returns a boolean that indicates if a software has already been installed
+   */
+  public static async isInstalled(software : string) : Promise<boolean> {
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  /**
   * Attempts to confirm a software's installation status
   * 
   * @param software is the software that is being checked
   * @returns a boolean that indicates if a software has already been installed
   */
  public static async isInstalled(software : string) : Promise<boolean> {
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

**Review by eetupur — APPROVED** [review]

File: src/classes/install-utils.ts [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * Install Minikube command
+   * 
+   * @returns Array of installation commands that are run by the wizard
+   */
+  public static async installKube(): Promise<string[]> {
+    const installUtil: string = await OsUtils.getCommand(CommandNames.installUtil);
+    const installRef: string = await InstallSwRefs.getInstallRef(installUtil, Software.Minikube);
+    const cmds : string[] = [];
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
    const cmds: string[] = [];
```

File: src/classes/install-utils.ts [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+  /**
+   * Install Kubernetes CLI command
+   * 
+   * @returns Array of installation commands that are run by the wizard
+   */
+  public static async installKubeCtl(): Promise<string[]> {
+    const installUtil: string = await OsUtils.getCommand(CommandNames.installUtil);
+    const installRef: string = await InstallSwRefs.getInstallRef(installUtil, Software.Minikube);
+    const cmds : string[] = [];
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
    const cmds: string[] = [];
```

File: src/classes/install-utils.ts [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    return (software == Software.Kustomize) ? installRef : `${installUtil} install ${installRef}`;
+  }
+
+  /**
+   * Attempts to confirm a software's installation status
+   * 
+   * @param software is the software that is being checked
+   * @returns a boolean that indicates if a software has already been installed
+   */
+  public static async isInstalled(software : string) : Promise<boolean> {
```
Comment (jarinykanen) [inline src/classes/install-utils.ts]:
```suggestion
  public static async isInstalled(software: string): Promise<boolean> {
```

**Review by jarinykanen — CHANGES_REQUESTED** [review]

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**